transaction/std/src/memo_builder/mod.rs

```diff
@@ -30,20 +30,20 @@ pub trait MemoBuilder: Debug {
     /// and gets a chance to report an error, if the fee is too large, or if it
     /// is being changed too late
     /// in the process, and memos that are already written would be invalid.
-    fn set_fee(&mut self, value: u64) -> Result<(), NewMemoError>;
+    fn set_fee(&mut self, amount: Amount) -> Result<(), NewMemoError>;
```

Comment:
Maybe this is also part of my question below (https://github.com/mobilecoinfoundation/mobilecoin/pull/1827/files#r859983201), why the `memo_builder` also has responsibility for the fee management. That may be out of scope for this PR, but I'm trying to understand the ways that the `TransactionBuilder` can fail, and the introduction of the `memo_builder`'s fee management for the `token_id`-specified fee has introduced a way for the `TransactionBuilder` to fail on `new`, which I'm curious about.

Reply:
the memo builder has to know the fee, because it has to get written into the destination memo. if the memo builder has the wrong fee, then the RTH data will be wrong.

So like, i've been thinking about this more, and I think that what happened here is, there is a tension:

* The way the transaction builder is currently architected, we actually build all `TxOut`'s at the time of calling `add_output`.
* The simplest way to add the encrypted memo stuff was, make memo encryption happen as part of `TxOut::new`. So that means each memo is finalized at time of `add_output` call, and not at the later `build` call.
* The memo builder is responsible for writing those memos. There are some "bad things" that can happen like, you could write the change output, and then try to change the fee. The way it works now is, the memo builder flags an error that gets returned by the `set_fee` call, and prevents the fee from changing and tells the user "you can't do that".

We could try to change it all so that, `TxOut::new` is not actually called until `.build()` is called, and instead we just store the data for `TxOut::new`. We could then give the memo builder all the final data at once -- all the inputs and `TxOut` amounts, the final fee value, etc. etc., and it could write all the memos for everything, which would simplify the memo builder.

I didn't do it that way before because it wasn't the simplest thing that could work, it would have been a bigger change to the `TransactionBuilder`. But we might be justified in doing a refactor like that, especially if the memo stuff gets more complicated, because that would probably remove the need for the memo builder to be stateful and update its state as outputs are written.

Reply:
Some random thoughts:
1) Completely agree on the tension between when memos are created and when TxOuts are created. I think there is no clear winner.
2) It is nice that `add_output` returns the `TxOut` - for example you get to store the public key, in case you want to reference it later in association with whatever you did that made you call `add_output`. This could potentially be resolved by returning some `TxOutContext` or something like that.
3) I think the way we are handling the fee right now is just incorrect - every transaction built must have a fee, and defaulting to `Mob::MINIMUM_FEE` is a hack. Now that `TransactionBuilder::new` accepts a fee, why not remove `set_fee()`. That way the fee cannot change after initialization, which I think is perfectly fine.

Reply:
(2) is a very good point, and I think in the future we will want even more stuff returned, for instance, there's this stuff about having a key that you can use to sign in a way that proves that you are the creator of a TxOut, that's in MCIP #31. And if we did the refactor I'm proposing then that would all get harder.
(3) The only use I can see for this is like, currently to get the minimum fees, clients make a network call to consensus. It might be helpful to the phones if they can make this call in parallel to building the transaction, since it takes like a second or so to actually build the transaction? and then they could update the fee only at the very end right before the change memo. Not sure that actually makes sense in real life, I think `.add_output` is really fast and only the `.build` step is slow.

We could make it so that, `TxOut::new` is called at `.add_output`, and we still return the confirmation number (and even the public key and stuff?), but in `.build`, the TransactionBuilder rewrites all the memos after everything about the transaction is known. But it's kind of messy.

Maybe the simplest thing is what you say, to just get rid of the `set_fee` option.

Reply:
Replacing the memos after we return a `TxOut` feels questionable to me. If someone later compares or hashes the entire `TxOut` they are going to have a bad time...

My vote is to remove the `set_fee` for now. As to the point of parallelizing, I think you could get the fee at the same time you are getting rings. Like you said, `add_output` is fast so waiting with the `add_output` calls doesn't seem problematic to me. And you can't add inputs until you have fetched rings.

Reply:
> and the introduction of the memo_builder's fee management for the token_id-specified fee has introduced a way for the TransactionBuilder to fail on new, which I'm curious about.

@sugargoat the reason that set-fee is allowed to fail is, when we specified RTH memos, we decided that i was okay to only reserve 6 bytes for the fee amount, since it seems wildly unlikely that higher fees than that would ever be paid in real life. So the memo builder gets to say "Sorry, this fee is too large to be represented. You can still use the transaction builder if you want but you have to turn off RTH or use a different memo builder or something."

If set-fee is now moving into the `new` call, then this failure gets propagated from there.

Reply:
> My vote is to remove the set_fee for now. As to the point of parallelizing, I think you could get the fee at the same time you are getting rings. Like you said, add_output is fast so waiting with the add_output calls doesn't seem problematic to me. And you can't add inputs until you have fetched rings.

Okay, let's make a ticket to explore removing the set fee API. I'm not sure if any client is using it, but we can find out